File: TalentShow/app/src/main/java/com/avatar/ava/presentation/main/fragments/semifinalists/SemifinalistsPresenter.java
package com.avatar.ava.presentation.main.fragments.semifinalists;

import android.content.Context;

import com.arellomobile.mvp.InjectViewState;
import com.arellomobile.mvp.MvpPresenter;
import com.avatar.ava.domain.Interactor;

import javax.inject.Inject;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

@InjectViewState
public class SemifinalistsPresenter extends MvpPresenter<SemifinalistsView> {

    private Interactor interactor;

    @Inject
    SemifinalistsPresenter(Interactor interactor) {
        this.interactor = interactor;
    }

    /*void getSemifinalists(){
        Disposable disposable = interactor.getSemifinalists()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(arrayList -> {
                            //getViewState().hideProgressBar();
                            getViewState().setSemifinalists(arrayList);
                        },
                        error -> {
                            //getViewState().hideProgressBar();
                            getViewState().showMessage("Не удалось загрузить рейтинг. Попробуйте позже");
                        });
    }*/

    void getProfile(int id){
        Disposable disposable = interactor.getPublicProfile(id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(person -> {
                            getViewState().setDataProfile(person);
                        },
                        error -> getViewState().showMessage("Упс. Произошла ошибка, попробуйте позже"));
    }

    void getBattles(){
        Disposable disposable = interactor.getBattles()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(arrayList -> {
                            //getViewState().hideProgressBar();
                            //getViewState().setSemifinalists(arrayList);
                            getViewState().setBattles(arrayList);
                        },
                        error -> {
                            //getViewState().hideProgressBar();
                            getViewState().showMessage("Не удалось загрузить рейтинг. Попробуйте позже");
                        });
    }

    void vote(String battleId, String semifinalistId){
        Disposable disposable = interactor.vote(battleId, semifinalistId)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(()-> {
                            //getViewState().hideProgressBar();
                            //getViewState().setSemifinalists(arrayList);
                            //getViewState().setBattles(arrayList);
                        },
                        error -> {
                            //getViewState().hideProgressBar();
                            getViewState().showMessage("Не удалось загрузить рейтинг. Попробуйте позже");
                        });
    }

    void cancelVote(String battleId, String semifinalistId){
        Disposable disposable = interactor.cancelVote(battleId, semifinalistId)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(()-> {
                            //getViewState().hideProgressBar();
                            //getViewState().setSemifinalists(arrayList);
                            //getViewState().setBattles(arrayList);
                        },
                        error -> {
                            //getViewState().hideProgressBar();
                            getViewState().showMessage("Не удалось загрузить рейтинг. Попробуйте позже");
                        });
    }
}

File: TalentShow/app/src/main/java/com/avatar/ava/presentation/main/fragments/battles/semifinalists/SemifinalistsPresenter.kt
package com.avatar.ava.presentation.main.fragments.battles.semifinalists

import com.arellomobile.mvp.InjectViewState
import com.arellomobile.mvp.MvpPresenter
import com.avatar.ava.domain.Interactor
import com.avatar.ava.domain.entities.semifinalists.BattleDTO
import com.avatar.ava.domain.entities.semifinalists.BattleParticipant
import com.avatar.ava.domain.entities.semifinalists.Semifinalist
import com.avatar.ava.presentation.main.fragments.battles.item.*
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import javax.inject.Inject

@InjectViewState
class SemifinalistsPresenter @Inject constructor(private val interactor: Interactor)
    : MvpPresenter<SemifinalistsView>() {

    private lateinit var compositeDisposable: CompositeDisposable

    override fun attachView(view: SemifinalistsView?) {
        super.attachView(view)
        compositeDisposable = CompositeDisposable()
        loadBattles()
    }

    fun loadBattles() {
        compositeDisposable.add(interactor.battles
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnSuccess {
                    viewState.enableUI()
                    it?.let { battles ->
                        val battleItems = battles.filterNotNull().mapToBattleItems()
                        viewState.setBattles(battleItems)
                    }
                }
                .doOnError {
                    viewState.disableUI("Не удалось загрузить рейтинг. Попробуйте позже")
                }
                .subscribe { _, _ -> })
    }

    fun vote(battleId: Int, battleParticipant: ParticipantItem) {
        compositeDisposable.add(interactor.vote(battleId, battleParticipant.semifinalist.id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnSuccess {
                    val votedSemifinalist =
                            battleParticipant.semifinalist.copy(
                                isLikedByUser = it.isLikedByUser,
                                votesNumber = it.votesNumber
                            )
                    val votedParticipant =
                            battleParticipant.copy(semifinalist = votedSemifinalist)

                    viewState.vote(battleId, votedParticipant)
                }
                .doOnError {
                    viewState.disableUI("Не удалось загрузить рейтинг. Попробуйте позже")
                }
                .subscribe { _, _ -> })
    }

    override fun detachView(view: SemifinalistsView?) {
        compositeDisposable.dispose()
        super.detachView(view)
    }

    private fun List<BattleParticipant>.mapToParticipantItems() : List<ParticipantItem> {
        return this.map {
            ParticipantItem(
                    semifinalist = it.semifinalist.toParticipantInfo(),
                    id = it.id,
                    name = it.name,
                    description = it.description,
                    profilePhoto = it.profilePhoto
            )
        }
    }

    private fun Semifinalist.toParticipantInfo() : SemifinalistItem {
        return SemifinalistItem(
                id = this.id,
                videoName = this.videoName,
                votesNumber = this.votesNumber,
                isLikedByUser = this.isLikedByUser
        )
    }

    private fun List<BattleDTO>.mapToBattleItems() : List<BattleItem> {
        return this.map {
            BattleItem(
                    id = it.id,
                    endDate = it.secondsUntilTheEnd,
                    votesNumber = it.totalVotesNumber,
                    winnersNumber = it.winnerNumber,
                    battleParticipants = it.battleParticipants.mapToParticipantItems().toMutableList()
            )
        }
    }
}